systemd_ctypes/bus.py

```diff
@@ -331,162 +332,491 @@ def add_object(self, path, obj):
-                                                   in_types, *args)
-            self._bus.send(message, None)
-        return invoker
+    This class provides high-level APIs for defining methods, properties, and
+    signals, as well as implementing introspection.
 
-    @staticmethod
-    def _setter_wrapper(obj, member, sig):
-        def invoker(self, value):
-            obj(self, value)
-            message = self._bus.message_new_signal(self._path, "org.freedesktop.DBus.Properties", "PropertiesChanged",
-                                                   'sa{sv}as',
-                                                   self._dbus_interface,
-                                                   { member: { 't': sig, 'v': value } },
-                                                   [])
-            self._bus.send(message, None)
-        return invoker
+    This class works well with these helper classes:
+      - Object.Method
+      - Object.Property
+      - Object.Signal
+      - Object.Interface
 
-    @staticmethod
-    def _add_info(category, member, *args):
-        def wrapper(obj):
-            mem = member
-            if mem is None:
-                mem = ''.join(part.title() for part in obj.__name__.split('_'))
-            if category == 'properties':
-                (sig,) = args
-                obj.setter = lambda obj: Object._setter_wrapper(obj, mem, sig)
-            elif category == 'signals':
-                (in_types,) = args
-                obj = Object._signal_wrapper(obj, mem, in_types)
-            obj._dbus_info = category, mem, args
-            return obj
-        return wrapper
+    An example class might look like:
 
-    @staticmethod
-    def interface(name):
-        return lambda cls: cls.prepare(name)
+        @bus.Object.Interface('com.example.MyObject')
+        class MyObject(bus.Object):
+            created = bus.Object.Signal('s', 'i')
+            renames = bus.Object.Property('u', value=0)
+            name = bus.Object.Property('s', 'undefined')
 
-    @staticmethod
-    def property(dbus_type, member=None):
-        return Object._add_info('properties', member, dbus_type)
+            @bus.Object.Method(out_types=(), in_types='s')
+            def rename(self, name):
+                self.renames += 1
+                self.name = name
 
-    @staticmethod
-    def signal(in_types=[], member=None):
-        return Object._add_info('signals', member, in_types)
+            def registered(self):
+                self.created('Hello', 42)
 
-    @staticmethod
-    def method(out_types=(), in_types=(), member=None):
-        return Object._add_info('methods', member, out_types, in_types)
+    See the documentation for the Method, Property, Signal and Interface
+    classes for more information.
+    """
+
+    # Class variables
+    _dbus_members: Optional[dict[str, dict[str, dict[str, dict[str, Any]]]]] = None
```

Comment:
Could this be not Optional? `interface` already relies on it not being `None`, no?